internal/magic/zip.go

```diff
@@ -55,26 +57,53 @@ type zipTokenizer struct {
 // next returns the next file name from the zip headers.
 // https://web.archive.org/web/20191129114319/https://users.cs.jmu.edu/buchhofp/forensics/formats/pkzip.html
 func (t *zipTokenizer) next() (fileName string) {
-	if t.i > len(t.in) {
-		return
+	// When the rest length is smaller than the header size, exit.
+	if len(t.in)-t.i < 30 {
+		return ""
 	}
+
 	in := t.in[t.i:]
-	// pkSig is the signature of the zip local file header.
-	pkSig := []byte("PK\003\004")
-	pkIndex := bytes.Index(in, pkSig)
-	// 30 is the offset of the file name in the header.
-	fNameOffset := pkIndex + 30
-	// end if signature not found or file name offset outside of file.
-	if pkIndex == -1 || fNameOffset > len(in) {
-		return
-	}
 
-	fNameLen := int(binary.LittleEndian.Uint16(in[pkIndex+26 : pkIndex+28]))
-	if fNameLen <= 0 || fNameOffset+fNameLen > len(in) {
-		return
+	offset := 0
+
+	// Read the first 4 bytes and look for the file header signature.
+	// If it is not at the start of buf, then set the current index to
+	// the first occurrence of the first byte of the signature in buf
+	// and re-run.
+	buf := in[offset : offset+4]
+	offset += 4
+	if !bytes.Equal(buf, zipHeader) {
+		i := bytes.IndexByte(buf, zipHeader[0])
```

Comment:
I don't understand why this if condition is needed. Can this be written without calling `bytes.IndexByte`? Basically go to known, predefined offsets without searching through `buf`.

Reply:
The intention behind this implementation was to support input buffers which do not start directly with the file header signature. But yeah, reading it window-wise in a 4 byte buffer is actually really silly when having the full input buffer at hand.

My original test implementation was with a Reader interface as `in` where this was sort of required. Now, the signature is searched in the `in` buffer directly.